dwave/cloud/cli.py

```diff
@@ -509,3 +517,138 @@ def upload(config_file, profile, problem_id, format, input_file):
+        return
+
+    # check all packages are registered/available
+    for pkg in packages:
+        if pkg not in contrib:
+            click.echo("Package {!r} not found.".format(pkg))
+            return 1
+
+    for pkg in packages:
+        _install_contrib_package(pkg, verbose=verbose)
+
+
+def _is_pip_package_installed(requirement):
+    """Checks if a pip `requirement` is installed."""
+
+    reqs = list(pkg_resources.parse_requirements(requirement))
+    assert len(reqs) == 1
+    req = reqs[0]
+
+    # NOTE: py35+ required
+    res = subprocess.run(
+        [sys.executable, "-m", "pip", "show", req.name],
+        stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+    return res.returncode == 0
+
+
+def _install_contrib_package(name, verbose=False):
+    """pip install non-oss package `name` from dwave's pypi repo."""
+
+    contrib = get_contrib_packages()
+    dwave_contrib_repo = "https://pypi.dwavesys.com/simple"
+
+    assert name in contrib
+    pkg = contrib[name]
+    title = pkg['title']
+
+    # check is `name` package is already installed
+    # right now the only way to check that is to check if all dependants from
+    # requirements are installed
+    # NOTE: currently we do not check if versions match!
+    if all(_is_pip_package_installed(req) for req in pkg['requirements']):
+        click.echo("{} already installed.".format(title))
+        return
+
+    # basic pkg info
+    click.echo(title)
+    click.echo(pkg['description'])
+
+    # license prompt
+    license = pkg['license']
+    msgtpl = ("This package is available under the {name!r} license.\n"
+              "The terms of the license are available online: {url!s}")
+    click.echo(msgtpl.format(name=license['name'], url=license['url']))
+
+    val = default_text_input('Install (y/n)?', default='y', optional=False)
+    if val.lower() != 'y':
+        click.echo('Skipping: {}'.format(title))
+        return
+
+    click.echo('Installing: {}'.format(title))
+    for req in pkg['requirements']:
+
+        # NOTE: py35+ required
+        res = subprocess.run(
+            [sys.executable, "-m", "pip", "install", req,
+             "--extra-index", dwave_contrib_repo],
+            stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+        if res.returncode or verbose:
+            click.echo(res.stdout)
+
+    click.echo('Successfully installed {}\n'.format(title))
+
+
+@cli.command()
+@click.option('--all', '-a', 'install_all', default=False, is_flag=True,
+              help='Install all contrib (non-OSS) packages')
+@click.option('--verbose', '-v', default=False, is_flag=True,
+              help='Increase output verbosity')
+def setup(install_all, verbose):
+    """Setup optional Ocean packages and configuration file(s)."""
+
+    if install_all:
+        click.echo("Installing all optional non-open-source packages.")
+    else:
+        prompt = "Do you want to install optional non-open-source packages (y/n)?"
```

Comment:
Maybe this should say something like "you will be prompted for each package individually"